### ISSUE TITLE
`size` for non-matrix arrays is slow and allocates memory

### ISSUE THREAD

**Issue by simonster** [JuliaLang/julia#9812]
``` julia
function f(a, b)
    for i = 1:1000000
        size(a) == size(b)
    end
end

julia> @time f([1], [1])
elapsed time: 0.096288154 seconds (48000208 bytes allocated, 47.00% gc time)
```

This does not affect tuples in general:

``` julia
function f(a, b)
    for i = 1:1000000
        a == b
    end
end

julia> @time f((1,), (1,))
elapsed time: 0.002870282 seconds (128 bytes allocated)
```

This hurts `==` for arrays, which checks equality of sizes before proceeding, as well as in-place functions that need to check that the output is the same size as the input.

**Comment by vtjnash** [comment]
duplicate of #9194

**Comment by vtjnash** [comment]
wait, no. it's that codegen only has a special-case for jl_f_arrayref of two arguments at https://github.com/JuliaLang/julia/blob/4b20e105929704877fc3a6e3a2518194e58cf3a0/src/codegen.cpp#L2120

```

julia> @code_typed f([1], [1])
1-element Array{Any,1}:
 :($(Expr(:lambda, Any[:a,:b], Any[Any[symbol("#s2"),symbol("#s1"),symbol("#s3"),:i,:_var2,:_var0,:_var1],Any[Any[:a,Array{Int64,1},0],Any[:b,Array{Int64,1},0],Any[symbol("#s2"),UnitRange{Int64},18],Any[symbol("#s1"),Int64,2],Any[symbol("#s3"),(Int64,Int64),18],Any[:i,Int64,18],Any[:_var2,(Int64,),18],Any[:_var0,Int64,18],Any[:_var1,Int64,18]],Any[]], :(begin  # none, line 2:
        #s2 = $(Expr(:new, UnitRange{Int64}, 1, :(((top(getfield))(Intrinsics,:select_value))((top(sle_int))(1,1000000)::Bool,1000000,(top(box))(Int64,(top(sub_int))(1,1)))::Int64)))
        #s1 = (top(getfield))(#s2::UnitRange{Int64},:start)::Int64
        unless (top(box))(Bool,(top(not_int))(#s1::Int64 === (top(box))(Int64,(top(add_int))((top(getfield))(#s2::UnitRange{Int64},:stop)::Int64,1))::Bool)) goto 1
        2: 
        _var0 = #s1::Int64
        _var1 = (top(box))(Int64,(top(add_int))(#s1::Int64,1))
        i = _var0::Int64
        #s1 = _var1::Int64 # line 3:
        _var2 = (top(arraysize))(a::Array{Int64,1})::(Int64,)
        _var2::(Int64,) == (top(arraysize))(b::Array{Int64,1})::(Int64,)::Bool
        3: 
        unless (top(box))(Bool,(top(not_int))((top(box))(Bool,(top(not_int))(#s1::Int64 === (top(box))(Int64,(top(add_int))((top(getfield))(#s2::UnitRange{Int64},:stop)::Int64,1))::Bool)))) goto 2
        1: 
        0: 
        return
    end::Void))))
```

**Comment by simonster** [comment]
Huh. It seems like we have `size(a::Matrix) = (arraysize(a,1), arraysize(a,2))` defined in `array.jl`, but for any other number of dimensions, `size` falls back to the unoptimized built-in.

**Comment by jakebolewski** [comment]
Couldn't `size{T,D}(a::Array{T,D})` be a staged function which would call `arraysize` similar to `Matrix`?

**Comment by JeffBezanson** [comment]
This now uses O(1) memory and is mostly fixed, except hits a performance cliff at ndims==5.

**Comment by KristofferC** [comment]
Can this be closed then?

**Comment by timholy** [comment]
Largely fixed in #16553. If it's still a problem then forced-inlining might improve things even more. (Some day, though, hopefully we won't be forced to use inlining just to handle varargs in a performant manner...)
